apps/src/tutorialExplorer/tutorialDetail.jsx

```diff
@@ -42,6 +50,44 @@ export default class TutorialDetail extends React.Component {
     }
   };
 
+  focusIn = event => {
+    // for starters, if our scroll has changed since the dialog opened, then reset it to where we were at
+    // the beginning. Otherwise, if you shift-tab above the detail on the first tutorial or tab below the detail
+    // on a tutorial towards the end of the window, it'll fidget the scroll. This keeps it consistent.
+    if (window.scrollY !== this.startingScroll[1]) {
+      window.scrollTo(...this.startingScroll);
+    }
+
+    /*
+      We've kept a ref to our dialog container, so when the user is adjusting focus we look to see if we're
+      focusing onto an element that's still in the modal. If so, then awesome! We're done. If not, then we look
+      in our modal and find the first button (which is the X to close) or the last link (which could vary based upon
+      what's in there).
+
+      If we're shifting focus away from the first button (and have left the modal), then we wrap around to the last link.
+      Otherwise, if we're leaving, we return to the start at the first button.
+
+      This feels very clunky. :-(
+
+      But it's to ensure that the user can only tab within the modal while it is open and cannot shift focus outside.
+
+     */
+    if (this.containerRef.current) {
+      const inModal = this.containerRef.current.contains(event.target);
+      if (!inModal) {
+        const firstButton =
+          this.containerRef.current.getElementsByTagName('button')[0];
+        const links = this.containerRef.current.getElementsByTagName('a');
```

Comment:
I'm concerned about the fragility of this section. There are a number of reasons we might add a button or another tab-able element to the end of this dialog, which would break this functionality. This leans too far on the side of hard-coding a fix, and in this case, I am going to say this comment is blocking, change required.

Now that aside, this does bring up a deeper question of how much do we prioritize a fix vs future maintainability. In your PR description, you brought up a number of changes that you recommend, but are too big for this task. For example, you mentioned using a component library dialog that addresses this issue more holistically. Normally our approach in these cases is to improve things as much we can while also keeping the code and user experience both at good quality. Then make a task to follow up with a bigger, holistic fix.

In this case, we do have a [component library dialog](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/componentLibrary/StylizedBaseDialog.jsx) that would benefit from what you're describing. That leaves two choices, either of which I'm happy to support and make space for in sprint planning.
1. Slim this PR down so it only changes the code in ways you're happy about, even if that means the user experience isn't perfect yet. Then make a task that covers the work to get us the rest of the way there.
2. Make a pre-factor PR that adds correct tab navigation to the component library dialog. Then migrate this to use that dialog (those two pieces could probably happen in either order).

Either is a good option and there are likely other good options too. Happy to chat more too!

Reply:
That's fair. I re-tooled it to use a more robust selector that'll catch all focusable elements on either end of the dialog.

 As a side note, I just added that selector text into this file, but it's a candidate to move to a separate library. OTOH, structurally we should largely discourage directly querying the dom in favor of staying in the react world, so keeping it in here to refactor out of existence later could be a better approach.

I would say that this PR is about as slimmed down as it can get to still properly address the issues, and that it is pretty surgical and self contained (it adds a new function, some additional set up (most of which is in self-contained new lifecycle methods), and the code to scroll element into view (which is also a single block).

As an alternative approach, the tiniest possible fix is to add an `autoFocus` attribute onto the `X` button of the dialog, but it doesn't _really_ work (the user could still tab off of the dialog (above or below), it steals focus from the tutorial list and never returns it, it doesn't address re-selection from navigation inside the dialog, etc). So this is the compromise slimmed down version to better address the request w/o causing more problems.

Swapping in a new dialog is a plausible option that'd help, but that alone wouldn't be enough to address the problems with changing tutorial focus from within the dialog via the area keys. And unfortunately that dialog component also has issues with focus falling off the dialog itself. So that dialog _also_ requires the same modifications added to the tutorial detail here. Not saying that it shouldn't be done there as well - just that the feature is scope creeping rapidly.

There are some other issues and inefficiencies in the tutorial page that we could address with a bigger scale refactor (such as the tutorial list constantly re-rendering), but again - scope creep way beyond the initial request.

Here's what I'd suggest:

1. Leave this PR (mostly?) as is (once everyone's confident in its behavior. I did address the noted ones already (thanks for the catches!)).
2. Re-tool the shared component dialog to ensure focus doesn't fall off of it (which is basically moving the `focusing` mods from here to that dialog), but maybe we should even skip over that because of note 3 here.
3. Then we can consider refactoring `tutorialDetail` to use the shared dialog, but honestly - we should also have a discussion about using a shared component library. `StylizedBasedDialog` appears to be mostly 2 years old and `BaseDialog` appears to be mostly 6 years old (looking at the git blames), and they're just simply more code we need to maintain. If it's viable to look at a third party system (material, blueprint, ant-design, whatever), then we could swap in something we don't need to maintain and that's known to already be ADA compliant. But I wouldn't want to just drop in a component system willy nilly, and if we do it, it should be with an eye towards migrating everything to it eventually, and with some sort of plan in place to do it.
4. Finally, we can look into a larger refactor of the tutorial explorer to more modern react, deal with the render inefficiencies, pin the filter list, keep closer track of selections, get rid of jQuery, etc.

Also happy to chat more and hopefully bring up your confidence in the proposed mods or discuss alternatives, as well as discuss what else we may want to do with the page going forward (if anything).